src/Engine/GameEvents/TeleportEvent.java

```diff
@@ -0,0 +1,154 @@
+    KILL,
+    SWAP
+  }
+
+  public TeleportEvent(GameMap map, Unit u, XYCoord dest, AnimationStyle animStyle, CollisionOutcome crashResult)
+  {
+    unit = u;
+    unitStart = new XYCoord(unit.x, unit.y);
+    unitDestination = dest;
+    animationStyle = animStyle;
+    collisionOutcome = crashResult;
+    unitDies = false;
+    obstacleDies = false;
+  }
+
+  public TeleportEvent(GameMap map, Unit u, XYCoord dest, AnimationStyle animStyle)
+  {
+    this(map, u, dest, animStyle, CollisionOutcome.KILL);
+  }
+
+  public TeleportEvent(GameMap map, Unit u, XYCoord dest, CollisionOutcome crashResult)
+  {
+    this(map, u, dest, AnimationStyle.BLINK, crashResult);
+  }
+
+  public TeleportEvent(GameMap map, Unit u, XYCoord dest)
+  {
+    this(map, u, dest, AnimationStyle.BLINK, CollisionOutcome.KILL);
+  }
+
+  @Override
+  public GameAnimation getEventAnimation(MapView mapView)
+  {
+    return mapView.buildTeleportAnimation(unit, unitStart, unitDestination, obstacle, animationStyle, unitDies, obstacleDies);
+  }
+
+  @Override
+  public void sendToListener(GameEventListener listener)
+  {
+    listener.receiveTeleportEvent(unit, unitStart, unitDestination);
+  }
+
+  @Override
+  public void performEvent(MapMaster gameMap)
+  {
+    gameMap.removeUnit(unit); // First remove our guy. He's in the ether.
+
+    // Figure out if something's in the way, and what to do with it.
+    obstacle = gameMap.getLocation(unitDestination).getResident();
+    obstacleDies = false;
+    if( null != obstacle )
+    {
+      switch(collisionOutcome)
+      {
+        case KILL:
+          obstacleDies = true;
+          break;
+        case SWAP:
+          // Move him to where our guy started. If he can't live there, he dies.
+          if( gameMap.isLocationValid(unitStart) )
+          {
+            gameMap.moveUnit(obstacle, unitStart.xCoord, unitStart.yCoord);
+            if( !obstacle.model.propulsion.canTraverse(gameMap.getEnvironment(unitStart)) )
+            {
+              obstacleDies = true;
+            }
+          }
+          else obstacleDies = true;
+          break;
+      }
+    }
+
+    if( obstacleDies )
+    {
+      UnitDieEvent ude = new UnitDieEvent(obstacle);
+      ude.performEvent(gameMap);
+
+      // Poor sap died; Check if his CO lost the game.
+      if( obstacle.CO.units.size() == 0 )
+      {
+        new CommanderDefeatEvent(obstacle.CO).performEvent(gameMap);
+      }
+    }
+
+    // Put our guy where he belongs.
+    gameMap.moveUnit(unit, unitDestination.xCoord, unitDestination.yCoord);
+
+    // If our guy can't survive there, end him.
+    if( !unit.model.propulsion.canTraverse(gameMap.getEnvironment(unitDestination)) )
+    {
+      unitDies = true;
+      new UnitDieEvent(unit).performEvent(gameMap);
+
+      // Our unit died; check if we are defeated.
+      if( unit.CO.units.size() == 0 )
+      {
+        // CO is out of units. Too bad.
+        new CommanderDefeatEvent(unit.CO).performEvent(gameMap);
+      }
```

Comment:
I've been thinking about our scheme of generating all possible events immediately, and I feel like maybe win/loss conditions in particular should just be checked whenever the event queue becomes free.
What say you?

Reply:
That could be difficult with certain events. Like, _losing_ an HQ is a lose condition, but _not having_ an HQ is not a lose condition. Not sure how to divorce evaluating lose conditions from the events that caused them.

Reply:
Hmm. I suppose what we could do is have `GameInstance` listen for and record stuff like "CO x lost a unit", and clear that state on having the loss-condition-check function called?
This also puts that logic right next to our `GameRules` in case we ever want to support fiddling with that.

Reply:
Putting that logic in `GameRules` is probably the long-term solution, since that will allow us to support alternate win conditions for scenario missions, but I don't think we are there yet.